pkg/controller/pravegacluster/upgrade.go

```diff
@@ -108,9 +115,60 @@ func (r *ReconcilePravegaCluster) clearUpgradeStatus(p *pravegav1alpha1.PravegaC
 	return nil
 }
 
-func (r *ReconcilePravegaCluster) syncComponentsVersion(p *pravegav1alpha1.PravegaCluster) (err error) {
-	var synced bool
+func (r *ReconcilePravegaCluster) rollbackClusterVersion(p *pravegav1alpha1.PravegaCluster, version string) (err error) {
+	_, rollbackCondition := p.Status.GetClusterCondition(pravegav1alpha1.ClusterConditionRollback)
+	if rollbackCondition == nil {
+		// We're in the first iteration for Rollback
+		// Add Rollback Condition to Cluster Status
+		p.Status.SetRollbackConditionTrue()
+		p.Spec.Version = version
```

Comment:
I am leery of changing the `PravegaCluster` spec; it is the user's goal, and is declaratively managed by configuration files near the installer.   The status can change to show that the upgrade was unsuccessful, but I think the user should trigger rollback by updating the goal.

Reply:
Agree that cluster `spec` represents the `desired` state and a controller(operator) should not change that value, since this is also the behavior exhibited by kubernetes when handling upgrades to deployments and statefulsets.

However, implementing automatic rollback from within the operator would require updating the cluster spec.
Automated rollback is what most customers would want in case of an upgrade failure as it removes the need for manual intervention and also makes sure the cluster stays in inconsistent state ( few pods will old and few with new version) for the shortest time.

A way to support both could be to have the operator do a manually triggered rollback  and automate invocation of this rollback when upgrade fails through some external means.
Please let me know your thoughts on this.

Reply:
I tried implementing Automated Rollback without changing `spec.Version` and rollback completed successfully, however post completion of rollback it triggers upgrade again, as the trigger for upgrade is `p.Spec.Version != p.Status.CurrentVersion`.

Reply:
Removed all occurrences of assignment to `p.Spec.Version` in operator code.
Implemented a manual rollback trigger when upgrade fails